Llampi

LaneRendell opened the pull request:
Refactoring Anthropic to take advantage/implement LLMAPI wrapper. Updated tests as well and tested. Added functionality to read ANTHROPIC API key from a .anthropic_key in same directory or user's home dir. Fall sback to env variable.

Commit:
Refactoring Anthropic to take advantage/implement LLMAPI wrapper. Updated tests as well and tested. Added functionality to read ANTHROPIC API key from a .anthropic_key in same directory or user's home dir. Fallsback to env variable.

Commit:
Just code cleanup, removing out come commented out code.

wbrown commented on api.go:
```diff
@@ -208,70 +260,14 @@ type Conversation struct {
 	HasThinkingContent bool
 }
 
-// NewConversation creates a new conversation with the given system prompt. It
```
Moving this to the bottom really impacts readability, this should be moved back to here.

wbrown commented on api.go:
```diff
@@ -504,6 +518,448 @@ func (conversation *Conversation) Send(text string) (
+// AddMessage adds a message to the conversation with the given role and
+// content. It used internally, and also can be used externally to
+// manipulate conversations.
+func (conversation *Conversation) AddMessage(role, content string) {
```
Your `AddMessage` implementation removes a lot of functionality that was there before.

* Images
* Tool use blocks
* Tool results
* Multiple content blocks

This is a significant regression for advanced use cases.

wbrown replied:
So, the signature here was:
`func (conversation *Conversation) AddMessage(role string, content *[]ContentBlock)`

But now it's:
`func (conversation *Conversation) AddMessage(role string, content string)`

Seems like we gave up a lot here, and broke the signature for prior users of `anthropic`.

wbrown commented on api.go:
```diff
@@ -504,6 +518,448 @@ func (conversation *Conversation) Send(text string) (
+			continue
+		}
+
+		switch currentEvent {
```
The parseSSEStream() only processes text deltas. Extended thinking content blocks (HasThinkingContent) won't be streamed.

wbrown commented on api.go:
```diff
@@ -504,6 +518,448 @@ func (conversation *Conversation) Send(text string) (
+// role/content strings for interface compliance.
+//
+// Note: Only the first text content block from each message is extracted.
+// Tool use, images, and other content types are not included in the output.
```
If the constraint is llmapi, then we should really extend llmapi to support the following:
* Multimodal content (images, documents)
* Tool use / function calling
* Tool results
* Thinking/reasoning blocks
* Multiple content blocks per message

wbrown reviewed (changes requested):
So, this makes me realize that we should really be extending llmapi so that we don't compromise on functionality here.

So, what's next is extending llmapi to support:
* Multimodal content (images, documents)
* Tool use / function calling
* Tool results
* Thinking/reasoning blocks
* Multiple content blocks per message

Keep in mind that my `anthropic` library is used by other projects that do use the functionality.

Commit:
Fixing issues from https://github.com/wbrown/anthropic/pull/2. Re-working code somewhat to implement new LLMAPI changes, and cleanup some things.

Commit:
Code cleanup for PR.

LaneRendell commented:
# PR: Extend llmapi for Rich Content Support

## Summary

This PR extends the `llmapi` interface to support multimodal content, tool use, tool results, thinking blocks, and multiple content blocks per message while maintaining backward compatibility.

## Changes

### llmapi Package

**New Types:**
- `Role` - Typed string for message roles (`RoleUser`, `RoleAssistant`, `RoleSystem`)
- `MediaType` - Typed string for MIME types (`MediaTypePNG`, `MediaTypeJPEG`, `MediaTypeGIF`, `MediaTypeWebP`, `MediaTypePDF`)
- `ContentBlock` - Unified content block supporting text, images, tool use, tool results, thinking, and documents
- `RichMessage` - Message with multiple content blocks
- `RichResponse` - Response with content blocks and helper methods (`Text()`, `ThinkingText()`, `ToolUses()`, `HasToolUse()`)
- `ToolDefinition` - Tool schema definition
- `Capabilities` - Provider capability flags

**New Interface Methods:**
- `SendRich(content []ContentBlock, sampling Sampling) (*RichResponse, error)`
- `SendRichStreaming(content []ContentBlock, sampling Sampling, callback StreamCallback) (*RichResponse, error)`
- `AddRichMessage(role string, content []ContentBlock)`
- `GetRichMessages() []RichMessage`
- `SetTools(tools []ToolDefinition)`
- `GetTools() []ToolDefinition`

**New Optional Interface:**
- `CapabilityProvider` with `GetCapabilities() Capabilities`

**Helper Constructors:**
- `NewTextBlock(text string)`
- `NewImageBlock(mediaType MediaType, base64Data string)`
- `NewImageBlockFromURL(mediaType MediaType, url string)`
- `NewToolResultBlock(toolUseID, content string, isError bool)`
- `NewThinkingBlock(thinking string)`

### anthropic Package

**Implemented:**
- All 6 new interface methods with full functionality
- `GetCapabilities()` returning Anthropic's supported features
- Conversion functions between llmapi and anthropic content block formats
- `parseSSEStreamRich()` for streaming with thinking block support
- Proper handling of `thinking_delta` SSE events

**Bug Fixes:**
- `SendRichStreaming` now properly sets up HTTP request
- API errors now include response body for debugging
- HTTP request retry now handles `http.NewRequest` errors
- Fixed comment grammar and removed dead code

### novelai Package (Pending)

Graceful degradation implementation:
- Rich methods extract text content only
- Non-text blocks (images, tools) are silently ignored
- Tools stored but not used (NovelAI doesn't support tool use)

## Breaking Changes

None. All existing methods continue to work unchanged.

## Type Safety Improvements

The following fields now use typed strings instead of plain strings:

| Type | Field | Old | New |
|------|-------|-----|-----|
| `Message` | `Role` | `string` | `llmapi.Role` |
| `RichMessage` | `Role` | `string` | `llmapi.Role` |
| `ImageSource` | `MediaType` | `string` | `llmapi.MediaType` |
| `DocumentSource` | `MediaType` | `string` | `llmapi.MediaType` |

## Files Changed

| File | Changes |
|------|---------|
| `llmapi/types.go` | New types, constants, helper constructors |
| `llmapi/interface.go` | Extended Conversation interface |
| `llmapi/types_test.go` | Tests for new types and helpers |
| `anthropic/api.go` | Full rich content implementation |
| `anthropic/api_test.go` | Tests for rich content methods |

## Test Plan

- [x] All llmapi unit tests pass (12 tests)
- [x] All anthropic unit tests pass (19 tests)
- [x] SSE stream parsing tests for text and thinking blocks
- [x] Content block conversion tests (to/from anthropic format)
- [x] RichResponse helper method tests

wbrown commented on api.go:
```diff
@@ -497,13 +759,458 @@ func (conversation *Conversation) Send(text string) (
+}
+
+// API URIs
+var messagesURI = "https://api.anthropic.com/v1/messages"
```
Package constants should really be at the top of the file, impacts readability.

LaneRendell replied:
https://github.com/wbrown/anthropic/pull/2#discussion_r2632545514

My thought was they would use AddRichMessage whereas AddMessage would be for simple text messages like NovelAPI has, but I can see how that is redundant and we can just have AddMessage handle both.

wbrown replied:
:+1: OK.

wbrown commented on api.go:
```diff
@@ -497,13 +759,458 @@ func (conversation *Conversation) Send(text string) (
+//
+//	event: message_stop
+//	data: {"type":"message_stop"}
+func (conversation *Conversation) parseSSEStreamRich(body io.Reader, callback llmapi.StreamCallback) (
```
Does this have duplicate parsing logic as in `SendStreaming`?

LaneRendell replied:
There's not really duplicate parsing logic parse handles the event processing, there's some basic parsing to grab the user message if one is provided, the system prompt, and some cache maangement, but most of it is dedicated to building the http request.

sendStreaming actually calls parseSSSEStreamRich to handle the processing of the event streams.

wbrown replied:
OK. :+1:

wbrown reviewed (changes requested):
Great job here. Some changes to make. On the bright side:

We now have:
* comprehensive llmapi Implementation
* proper content block conversion
* thinking block support in streaming
* good test coverage
* API errors returning response body, improving debugging.

Great job.

Commit:
Fixing up something for code review.
